service/storage_service.cc

```diff
@@ -3209,5 +3283,68 @@ bool storage_service::is_repair_based_node_ops_enabled() {
     return _db.local().get_config().enable_repair_based_node_ops();
 }
 
+node_ops_meta_data::node_ops_meta_data(gms::inet_address coordinator, shared_ptr<node_ops_info> ops, std::function<future<> ()> func)
+    : _coordinator(std::move(coordinator))
+    , _abort(std::move(func))
+    , _ops(std::move(ops))
+    , _watchdog([this] { exec_watchdog(); }) {
+    _watchdog.arm(std::chrono::seconds(30));
+}
+
+future<> node_ops_meta_data::abort() {
+    if (_ops) {
+        _ops->abort = true;
+    }
+    _watchdog.cancel();
+    return _abort();
+}
+
+void node_ops_meta_data::update_watchdog() {
+    _watchdog.cancel();
+    _watchdog.arm(std::chrono::seconds(30));
+}
+
+void node_ops_meta_data::cancel_watchdog() {
+    _watchdog.cancel();
+}
+
+void node_ops_meta_data::exec_watchdog() {
+    abort().handle_exception([this] (std::exception_ptr ep) {
+        slogger.warn("Failed to run node ops abort for coordinator={}: {}", _coordinator, ep);
+    });
```

Comment:
This can also cause a use after free, if this is deallocated before handle_exception gets a chance to execute.